### ISSUE TITLE
Visual selection is not adjusted when changing text selection with TalkBack

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#22971]
To reproduce, run the flutter gallery text fields demo and enter some text. Toggle TalkBack on and open the local context menu to perform text selection.

Expected: changes in text selection are reflected in the text field

Actual: TalkBack announces the correct selection but text field visuals are unchanged.

**Comment by chunhtai** [comment]
It looks like this is not related the change i made,

I have seen this error message when performing selection

```
E/flutter ( 8209): [ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: type '_InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, int>'
E/flutter ( 8209): #0      SemanticsConfiguration.onSetSelection=.<anonymous closure> (package:flutter/src/semantics/semantics.dart:3035:30)
E/flutter ( 8209): #1      SemanticsOwner.performAction (package:flutter/src/semantics/semantics.dart:2561:14)
E/flutter ( 8209): #2      _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding._handleSemanticsAction (package:flutter/src/rendering/binding.dart:270:36)
E/flutter ( 8209): #3      _invoke3.<anonymous closure> (dart:ui/hooks.dart:264:15)
E/flutter ( 8209): #4      _rootRun (dart:async/zone.dart:1124:13)
E/flutter ( 8209): #5      _CustomZone.run (dart:async/zone.dart:1021:19)
E/flutter ( 8209): #6      _CustomZone.runGuarded (dart:async/zone.dart:923:7)
E/flutter ( 8209): #7      _invoke3 (dart:ui/hooks.dart:263:10)
E/flutter ( 8209): #8      _dispatchSemanticsAction (dart:ui/hooks.dart:160:3)
```

**Comment by dnfield** [comment]
@chunhtai can you try changing 

```dart
set onSetSelection(SetSelectionHandler value) {
    assert(value != null);
    _addAction(SemanticsAction.setSelection, (dynamic args) {
      final Map<String, int> selection = args;
      assert(selection != null && selection['base'] != null && selection['extent'] != null);
      value(TextSelection(
        baseOffset: selection['base'],
        extentOffset: selection['extent'],
      ));
    });
    _onSetSelection = value;
  }
```

to

```dart
set onSetSelection(SetSelectionHandler value) {
    assert(value != null);
    _addAction(SemanticsAction.setSelection, (dynamic args) {
      assert(args != null && args is Map);
      final Map<String, int> selection = args.cast<String, int>();
      assert(selection != null && selection['base'] != null && selection['extent'] != null);
      value(TextSelection(
        baseOffset: selection['base'],
        extentOffset: selection['extent'],
      ));
    });
    _onSetSelection = value;
  }
```

**Comment by chunhtai** [comment]
yes this is type conversion issue, I already verified it works with correct type conversion, i am trying to figure out which commit break this.

**Comment by dnfield** [comment]
Possible that this was a Dart change - I believe this syntax was allowed in Dart 1 but not Dart 2.

**Comment by chunhtai** [comment]
The functionality is implemented https://github.com/flutter/flutter/pull/14275 which should have broken at that time, the args that passes in is already a Ma<dynamic, dynamic>. The channel is implemented back in May 1st 2017 and hasn't changed

**Comment by chunhtai** [comment]
@dnfield yes, that sounds correct, I will have a pr up to do the correct conversion.

**Comment by dnfield** [comment]
I said syntax but meant runtime type conversion :)

This changed on the stable version of Dart 2.0, which I think was somewhere around early 2018.  We probably have other places in the framework that have this bug.  We can fix it turning on no_implicit_downcasts and properly fixing the callsites.  But there are a lot of them.
